internal/npm_install/generate_build_file.ts

```diff
@@ -0,0 +1,1140 @@
+ * exist
+ * @param {any} pkg
+ * @param {string} path
+ * @returns {string | undefined}
+ */
+function findEntryFile(pkg: Dep, path: string) {
+  const cleanPath = cleanupEntryPointPath(path);
+  // check if main entry point exists
+  const entryFile = findFile(pkg, cleanPath) || findFile(pkg, `${cleanPath}.js`);
+  if (!entryFile) {
+    // If entryPoint entry point listed could not be resolved to a file
+    // This can happen
+    // in some npm packages that list an incorrect main such as v8-coverage@1.0.8
+    // which lists `"main": "index.js"` but that file does not exist.
+    log_verbose(
+        `could not find entry point for the path ${cleanPath} given by npm package ${pkg._name}`);
+  }
+  return entryFile;
+}
+
+/**
+ * Tries to resolve the entryPoint file from the pkg for a given mainFileName
+ *
+ * @param {any} pkg
+ * @param {'browser' | 'module' | 'main'} mainFileName
+ * @returns {string | undefined} the path or undefined if we cant resolve the file
+ */
+function resolveMainFile(pkg: Dep, mainFileName: string) {
+  const mainEntryField = pkg[mainFileName];
+
+  if (mainEntryField) {
+    if (typeof mainEntryField === 'string') {
+      return findEntryFile(pkg, mainEntryField)
+
+    } else if (typeof mainEntryField === 'object' && mainFileName === 'browser') {
+      // browser has a weird way of defining this
+      // the browser value is an object listing files to alias, usually pointing to a browser dir
+      const indexEntryPoint = mainEntryField['index.js'] || mainEntryField['./index.js'];
+      if (indexEntryPoint) {
+        return findEntryFile(pkg, indexEntryPoint)
+      }
+    }
+  }
+}
+
+/**
+ * Tries to resolve the mainFile from a given pkg
+ * This uses seveal mainFileNames in priority to find a correct usable file
+ * @param {any} pkg
+ * @returns {string | undefined}
+ */
+function resolvePkgMainFile(pkg: Dep) {
+  // es2015 is another option for mainFile here
+  // but its very uncommon and im not sure what priority it takes
+  //
+  // this list is ordered, we try resolve `browser` first, then `module` and finally fall back to
+  // `main`
+  const mainFileNames = ['browser', 'module', 'main']
+
+      for (const mainFile of mainFileNames) {
+    const resolvedMainFile = resolveMainFile(pkg, mainFile);
+    if (resolvedMainFile) {
+      return resolvedMainFile;
+    }
+  }
+
+  // if we cant find any correct file references from the pkg
+  // then we just try looking around for common patterns
+  const maybeRootIndex = findEntryFile(pkg, 'index.js');
+  if (maybeRootIndex) {
+    return maybeRootIndex
+  }
+
+  const maybeSelfNamedIndex = findEntryFile(pkg, `${pkg._name}.js`);
+  if (maybeSelfNamedIndex) {
+    return maybeSelfNamedIndex;
+  }
+
+  // none of the methods we tried resulted in a file
+  log_verbose(`could not find entry point for npm package ${pkg._name}`);
+
+  // at this point there's nothing left for us to try, so return nothing
+  return undefined;
+}
+
+type Bag<T> =
```

Comment:
this type is usually called a Dictionary or Dict, typescript has one called a `Record<T,V>` so you could default it like `type Dictionary<T> = Record<string,T>`

Reply:
I can empathize with that argument coming from python. I'm ambivalent on this because Dicts are not exactly javascript objects and Object is reserved in js for obvious reasons. I hadn't considered utility types. thanks